### PR TITLE
refactor(toml,msgpack,crypto,data-structures): align the error messages to the style guide

### PR DESCRIPTION
**Pull request by irbull**
Aligns the error messages for the following folders: `toml`, `msgpack`, `crypto` and `data_structures`.

### CODE REVIEW

File: data_structures/binary_heap.ts [hunk 1]
```diff
@@ -71,7 +71,7 @@ export class BinaryHeap<T> implements Iterable<T> {
   constructor(compare: (a: T, b: T) => number = descend) {
     if (typeof compare !== "function") {
       throw new TypeError(
-        "compare must be a function, did you mean to use BinaryHeap.from?",
+        "The 'compare' parameter must be a function, did you mean to use BinaryHeap.from?",
```
Comment (iuioiua) [inline data_structures/binary_heap.ts:74]:
```suggestion
        "The 'compare' parameter must be a function. Did you mean to use BinaryHeap.from?",
```

File: data_structures/binary_search_tree.ts [hunk 2]
```diff
@@ -107,7 +107,7 @@ export class BinarySearchTree<T> implements Iterable<T> {
   constructor(compare: (a: T, b: T) => number = ascend) {
     if (typeof compare !== "function") {
       throw new TypeError(
-        "compare must be a function, did you mean to call BinarySearchTree.from?",
+        "The 'compare' parameter must be a function, did you mean to call BinarySearchTree.from?",
```
Comment (iuioiua) [inline data_structures/binary_search_tree.ts:110]: Ditto
```suggestion
        "The 'compare' parameter must be a function. Did you mean to call BinarySearchTree.from?",
```

File: crypto/crypto.ts [hunk 3]
```diff
@@ -235,7 +235,9 @@ const stdCrypto: StdCrypto = ((x) => x)({
           for await (const chunk of data as AsyncIterable<BufferSource>) {
             const chunkBytes = toUint8Array(chunk);
             if (!chunkBytes) {
-              throw new TypeError("data contained chunk of the wrong type");
+              throw new TypeError(
+                "Data contained a chunk with an invalid type, chunks must be ArrayBuffer or Uint8Array",
```
Comment (iuioiua) [inline crypto/crypto.ts:239]: How's this? Ditto elsewhere.
```suggestion
                "Chunks must be instances of ArrayBuffer or Uint8Array",
```

File: data_structures/binary_heap.ts (hunk 1 above)
Comment (irbull) [reply data_structures/binary_heap.ts:74]: The rules we have indicate that a period should not be used within an error message. The idea is that we don't want several sentences, which is why I've been trying to use a single sentence (possibly with a colon). Do we want to change this rule?

Comment (kt3k) [reply data_structures/binary_heap.ts:74]: What do Web APIs, V8, Deno CLI do?

Comment (timreichen) [reply data_structures/binary_heap.ts:74]: I think V8 uses commas: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Cant_define_property_object_not_extensible

Comment (timreichen) [reply data_structures/binary_heap.ts:74]: Just found an example with colon: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Regex_nothing_to_repeat

File: crypto/crypto.ts (hunk 3 above)
Comment (kt3k) [reply crypto/crypto.ts:239]: > Message should state the action that lead to the error:

> Additional information should describe the current state:

```suggestion
                "Cannot digest the data: A chunk is not ArrayBuffer nor ArrayBufferView",
```

Comment (kt3k) [reply crypto/crypto.ts:239]: Note: `toUint8Array` seems accepting `ArrayBuffer` and `ArrayBufferView` https://github.com/denoland/std/blob/3c960730b7f2c3de4bf51ff7a87054bfbf03d477/crypto/crypto.ts#L152

File: crypto/crypto.ts [hunk 4]
```diff
@@ -270,7 +272,9 @@ const stdCrypto: StdCrypto = ((x) => x)({
         for (const chunk of data) {
           const chunkBytes = toUint8Array(chunk);
           if (!chunkBytes) {
-            throw new TypeError("data contained chunk of the wrong type");
+            throw new TypeError(
+              "Data contained a chunk with an invalid type, chunks must be ArrayBuffer or Uint8Array",
```
Comment (kt3k) [inline crypto/crypto.ts:276]: ditto
```suggestion
              "Cannot digest the data: A chunk is not ArrayBuffer nor ArrayBufferView",
```

File: data_structures/binary_search_tree.ts (hunk 2 above)
Comment (kt3k) [reply data_structures/binary_search_tree.ts:110]: I also suggest we should prepend `Cannot construct a BinarySearchTree:`

File: data_structures/red_black_tree.ts [hunk 5]
```diff
@@ -109,7 +109,7 @@ export class RedBlackTree<T> extends BinarySearchTree<T> {
   constructor(compare: (a: T, b: T) => number = ascend) {
     if (typeof compare !== "function") {
       throw new TypeError(
-        "compare must be a function, did you mean to call RedBlackTree.from?",
+        "The 'compare' parameter must be a function, did you mean to call RedBlackTree.from?",
```
Comment (kt3k) [inline data_structures/red_black_tree.ts:112]:
```suggestion
        "Cannot construct a RedBlackTree: The 'compare' parameter must be a function, did you mean to call RedBlackTree.from?",
```

File: msgpack/decode.ts [hunk 6]
```diff
@@ -82,7 +82,7 @@ function decodeMap(
 
     if (typeof key !== "number" && typeof key !== "string") {
       throw new EvalError(
-        "Messagepack decode came across an invalid type for a key of a map",
+        "Messagepack decode came across an invalid type for a key of a map, keys must be a number or a string",
```
Comment (kt3k) [inline msgpack/decode.ts:85]:
```suggestion
        "Cannot decode a key of a map: The type of key is invalid, keys must be a number or a string",
```

File: msgpack/decode.ts [hunk 7]
```diff
@@ -205,7 +205,7 @@ function decodeSlice(
     case 0xc7: // ext 8 - small extension type
     case 0xc8: // ext 16 - medium extension type
     case 0xc9: // ext 32 - large extension type
-      throw new Error("ext not implemented yet");
+      throw new Error("Large extension type 'ext' not implemented yet");
```
Comment (kt3k) [inline msgpack/decode.ts:208]:
```suggestion
      throw new Error("Cannot decode a slice: Large extension type 'ext' not implemented yet");
```

File: msgpack/decode.ts [hunk 8]
```diff
@@ -311,7 +311,7 @@ function decodeSlice(
     case 0xd6: // fixext 4 - 4 byte extension type
     case 0xd7: // fixext 8 - 8 byte extension type
     case 0xd8: // fixext 16 - 16 byte extension type
-      throw new Error("fixext not implemented yet");
+      throw new Error("'fixext' not implemented yet");
```
Comment (kt3k) [inline msgpack/decode.ts:314]:
```suggestion
      throw new Error("Cannot decode a slice: 'fixext' not implemented yet");
```

File: toml/_parser.ts [hunk 9]
```diff
@@ -99,7 +99,10 @@ export class Scanner {
     // Invalid if current char is other kinds of whitespace
     if (!this.isCurrentCharEOL() && /\s/.test(this.char())) {
       const escaped = "\\u" + this.char().charCodeAt(0).toString(16);
-      throw new SyntaxError(`Contains invalid whitespaces: \`${escaped}\``);
+      const position = this.#position;
+      throw new SyntaxError(
+        `Contains invalid whitespace at position '${position}': \`${escaped}\``,
```
Comment (kt3k) [inline toml/_parser.ts:104]:
```suggestion
        `Cannot parse the TOML: It contains invalid whitespace at position '${position}': \`${escaped}\``,
```

File: toml/_parser.ts [hunk 10]
```diff
@@ -153,7 +156,7 @@ export function deepAssignWithTable(target: Record<string, unknown>, table: {
   value: Record<string, unknown>;
 }) {
   if (table.key.length === 0 || table.key[0] == null) {
-    throw new Error("Unexpected key length");
+    throw new Error("Key length must be a positive number");
```
Comment (kt3k) [inline toml/_parser.ts:159]:
```suggestion
    throw new Error("Cannot parse the TOML: Key length must be a positive number");
```

Comment (timreichen) [reply toml/_parser.ts:159]: Shouldn't it be `Key length is not a positive number`?

File: data_structures/binary_heap.ts (hunk 1 above)
Comment (irbull) [reply data_structures/binary_heap.ts:74]: I think the original rule said either comma `,` or colon `:`, but not period. I like that rule as there are often times a comma makes sense, but we don't want several sentences (that's why no period).

I've updated a number of these error messages, and without the use of a comma, I think they get very hard to read.

File: data_structures/red_black_tree.ts (hunk 5 above)
Comment (timreichen) [reply data_structures/red_black_tree.ts:112]: I think this should also be formulated as the current state: "the compare parameter is not a function"

File: data_structures/binary_search_tree.ts (hunk 2 above)
Comment (irbull) [reply data_structures/binary_search_tree.ts:110]:
```suggestion
        "Cannot construct a BinarySearchTree: the 'compare' parameter must be a function: did you mean to call BinarySearchTree.from?",
```

File: data_structures/red_black_tree.ts (hunk 5 above)
Comment (irbull) [reply data_structures/red_black_tree.ts:112]:
```suggestion
        "Cannot construct a RedBlackTree: The 'compare' parameter is not a function, did you mean to call RedBlackTree.from?,
```

File: data_structures/red_black_tree_test.ts [hunk 11]
```diff
@@ -8,7 +8,7 @@ Deno.test("RedBlackTree throws if compare is not a function", () => {
   assertThrows(
     () => new RedBlackTree({} as (a: number, b: number) => number),
     TypeError,
-    "compare must be a function",
+    "The 'compare' parameter must be a function",
```
Comment (irbull) [inline data_structures/red_black_tree_test.ts:11]:
```suggestion
    "Cannot construct a RedBlackTree: The 'compare' parameter is not a function",
```

File: toml/_parser.ts (hunk 10 above)
Comment (irbull) [reply toml/_parser.ts:159]:
```suggestion
    throw new Error("Cannot parse the TOML: key length is not a positive number"));
```

File: toml/parse_test.ts [hunk 12]
```diff
@@ -515,7 +515,7 @@ Deno.test({
           },
         ),
       Error,
-      "Unexpected key length",
+      "Key length must be a positive number",
```
Comment (irbull) [inline toml/parse_test.ts:518]:
```suggestion
      "Cannot parse the TOML: key length is not a positive number",
```

File: data_structures/binary_heap.ts (hunk 1 above)
Comment (irbull) [reply data_structures/binary_heap.ts:74]:
```suggestion
        "Cannot construct a BinaryHeap: the 'compare' parameter is not a function, did you mean to call BinaryHeap.from?",
```

**Review by kt3k — APPROVED** [review]
LGTM

**Pull request merged**